flang/lib/Lower/CallInterface.cpp

```diff
@@ -486,3 +501,64 @@ Fortran::lower::CallInterface<T>::getResultType() const {
 
 template class Fortran::lower::CallInterface<Fortran::lower::CalleeInterface>;
 template class Fortran::lower::CallInterface<Fortran::lower::CallerInterface>;
+
+/// Build signature from characterstics when there is no Fortran entity to
+/// associate with the arguments (i.e, this is not a call site or a procedure
+/// declaration. This is needed when dealing with function pointers/dummy
+/// arguments.
+
+class SignatureBuilder;
+template <>
+struct Fortran::lower::PassedEntityTypes<SignatureBuilder> {
+  using FortranEntity = FakeEntity;
+  using FirValue = int;
+};
+
+class SignatureBuilder
+    : public Fortran::lower::CallInterface<SignatureBuilder> {
+public:
+  SignatureBuilder(const Fortran::evaluate::characteristics::Procedure &p,
+                   Fortran::lower::AbstractConverter &c, bool forceImplicit)
+      : CallInterface{c}, proc{p} {
+    if (forceImplicit || proc.CanBeCalledViaImplicitInterface())
+      buildImplicitInterface(proc);
+    else
+      buildExplicitInterface(proc);
+  }
+  bool hasAlternateReturns() const {
+    for (const auto &dummy : proc.dummyArguments)
+      if (std::holds_alternative<
+              Fortran::evaluate::characteristics::AlternateReturn>(dummy.u))
+        return true;
+    return false;
+  };
+  std::string getMangledName() const { return ""; }
```

Comment:
Should this be a hard failure? A null name is invalid in IR terms.

Reply:
Good point, changed into llvm::unreachable